lib/devicetrust/assert/assert.go

```diff
@@ -0,0 +1,180 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package assert
+
+import (
+	"context"
+
+	"github.com/gravitational/trace"
+	"google.golang.org/grpc"
+
+	devicepb "github.com/gravitational/teleport/api/gen/proto/go/teleport/devicetrust/v1"
+	"github.com/gravitational/teleport/lib/devicetrust/authn"
+)
+
+// AssertDeviceClientStream is the client-side device assertion stream.
+type AssertDeviceClientStream interface {
+	Send(*devicepb.AssertDeviceRequest) error
+	Recv() (*devicepb.AssertDeviceResponse, error)
+}
+
+// Ceremony implements the client-side assertion ceremony.
+//
+// See [devicepb.AssertDeviceRequest] for details.
+type Ceremony struct {
+	// newAuthnCeremony defaults to authn.NewCeremony.
+	newAuthnCeremony func() *authn.Ceremony
+}
+
+// CeremonyOpt is a creation option for [Ceremony].
+type CeremonyOpt func(*Ceremony)
+
+// WithNewAuthnCeremonyFunc overrides the default authn.Ceremony constructor,
+// allowing callers to change the underlying assert implementation.
+//
+// Useful for testing. Avoid for production code.
+func WithNewAuthnCeremonyFunc(f func() *authn.Ceremony) CeremonyOpt {
+	return func(c *Ceremony) {
+		c.newAuthnCeremony = f
+	}
+}
+
+// NewCeremony creates a new [Ceremony], binding all native device trust
+// methods.
+func NewCeremony(opts ...CeremonyOpt) (*Ceremony, error) {
+	c := &Ceremony{
+		newAuthnCeremony: authn.NewCeremony,
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	return c, nil
+}
+
+// Run runs the client-side device assertion ceremony.
+// Requests and responses are consumed from the stream until the device is
+// asserted or authentication fails.
+func (c *Ceremony) Run(ctx context.Context, stream AssertDeviceClientStream) error {
+	newAuthn := c.newAuthnCeremony
+	if newAuthn == nil {
+		newAuthn = authn.NewCeremony
+	}
+
+	devices := &devicesClientAdapter{
+		stream: stream,
+	}
+	certs := &devicepb.UserCertificates{} // required but unused
+
+	// Implement Assertion in terms of Authentication, so we borrow both the
+	// Secure Enclave and TPM branches from it.
+	// TODO(codingllama): Refactor so we don't need so many adapters?
+	_, err := newAuthn().Run(ctx, devices, certs)
+	return trace.Wrap(err)
+}
+
+type devicesClientAdapter struct {
+	devicepb.DeviceTrustServiceClient
+
+	stream AssertDeviceClientStream
+}
+
+func (d *devicesClientAdapter) AuthenticateDevice(ctx context.Context, opts ...grpc.CallOption) (devicepb.DeviceTrustService_AuthenticateDeviceClient, error) {
+	return &authnStreamAdapter{
+		ctx:    ctx,
+		stream: d.stream,
+	}, nil
+}
+
+type authnStreamAdapter struct {
+	devicepb.DeviceTrustService_AuthenticateDeviceClient
+
+	ctx    context.Context
```

Comment:
Do we need a context here? It seems the only needed context is already passed to the `Run` method.
Also, in some places we name this context as `closeContext` or similar, which I find more clear because it clearly distinguishes it from a method's context.

Reply:
We need this for the stream's Context() method. gRPC streams hold the context inside them, so if we want to provide an accurate behavior we have to do so too.

- https://pkg.go.dev/google.golang.org/grpc#ClientStream